examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+      end
+
+      def onKeyDown(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keydown(key, view, @mouse_ip, @picked_first_ip)
+        pick_mouse_position(view)
+      end
+
+      def onKeyUp(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keyup(key, view)
+        pick_mouse_position(view)
+      end
+
+      def onMouseMove(flags, x, y, view)
+        if @selection.empty?
+          # If the tool selection is empty, try picking the hovered object.
+          try_select_entity(view, x, y)
+        end
+
+        @mouse_position = Geom::Point3d.new(x, y, 0)
+        pick_mouse_position(view)
+      end
+
+      def onLButtonDown(flags, x, y, view)
+        @mouse_down = Geom::Point3d.new(x, y)
+
+        # If tool selection is empty, pick our hovered entity.
+        if @selection.empty?
+          @selection = view.model.selection.to_a
+        end
+
+        return if @selection.empty?
+
+        # For a Move or Rotate like tool, we can select a starting point in the
+        # same mouse click as we select an entity. For Scale or other tools
+        # where you need to select a handle, this logic is typically slightly
+        # different, and separate click is needed.
+
+        if !picked_first_point?
+          start_move(view)
+        else
+          end_move(view)
+        end
+
+        update_ui
+        view.invalidate
+      end
+
+      def onLButtonUp(flags, x, y, view)
+        if @mouse_down.distance([x, y]) > DRAG_THRESHOLD
+          end_move(view)
+        end
+      end
+
+      # Here we have hard coded a special ID for the move cursor in SketchUp.
+      # Normally you would use `UI.create_cursor(cursor_path, 0, 0)` instead
+      # with your own custom cursor bitmap:
+      #
+      #   CURSOR = UI.create_cursor(cursor_path, 0, 0)
+      MOVE_CURSOR = 641
+      INVALID_CURSOR = 663
+
+      def onSetCursor
+        # Note that `onSetCursor` is called frequently so you should not do much
+        # work here. At most you switch between different cursor representing
+        # the state of the tool.
+
+        # Cursor depends on the tool state and hovered entity.
+        # If the user is at the selection stage and hover a non-selectable
+        # entity, they se an invalid-cursor.
+        # Hovering empty space in SketchUp doesn't show the invalid-cursor,
+        # even if empty space can't be selected.
+        if !picked_first_point? && @hovered_entity && !@valid_hovered_entity
+          UI.set_cursor(INVALID_CURSOR)
+        else
+          UI.set_cursor(MOVE_CURSOR)
+        end
+      end
+
+      def onUserText(text, view)
+        begin
+          distance = text.to_l
+        rescue ArgumentError
+          UI.messagebox('Invalid length')
+          return
+        end
+        direction = picked_points[1] - picked_points[0]
+        end_point = picked_points[0].offset(direction, distance)
+        @mouse_ip = Sketchup::InputPoint.new(end_point)
+
+        do_move(view)
```

Comment:
I would like to see a comment here on what's going on. Seems like you are simulating mouse input?

It seems strange to me to generate an input point from text input though... This is one type of user input driving another form of user input to drive the state of the tool.

```
keyboard -> mouse -> tool-state
```

I'd prefer to see something like:

```
keyboard -+
          +-->tool-state
   mouse -+
```

Reply:
You could have a separate instance variable for a length and point and have logic to determine what value to use when, but I'm thinking it's simpler to just have a single value.

Reply:
Could you add some comment here to explain what is happening. I'm still not exactly clear on what we're doing where. Why we need to call `do_move` and `end_move` in immediate succession.

Could `end_move` that the mouse event calls call another function that performs the move, and this onUserText could then use directly?

Reply:
When you enter a value you also finish the operation, not just updating the current position. Adding comments to clarify.